Fix handling of device

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81113
* #78616
* __->__ #78615
* #78350
* #80219

Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

cc @ezyang @mruberry @ngimel @Lezcano @peterbell10

lezcano committed:
Fix handling of device

Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

mruberry commented on torch/_prims/utils.py:
```diff
@@ -75,7 +75,7 @@ def __new__(
         strides: Optional[StrideType] = None,
         dtype: Optional[torch.dtype] = None,
-        device: Optional[Union[torch.device, str]] = None,
+        device: Optional[DeviceLikeType] = None,
```
Nice

mruberry commented on torch/_prims/utils.py:
```diff
@@ -217,7 +208,7 @@ def check_significant_strides(
     # Only compares strides that are "meaningful" -- strides for dimensions with length > 1
     # and for tensors with more than one element
-    if (a.device.type == "cuda" or b.device.type == "cuda") and a.numel() > 0:
+    if (is_cuda(a.device) or is_cuda(b.device)) and a.numel() > 0:
```
I prefer querying the type directly vs a sugar here

mruberry commented on torch/_prims/utils.py:
```diff
@@ -489,14 +480,22 @@ def check_same_device(*args, allow_cpu_scalar_tensors):
 
 
-def canonicalize_device(device: Union[str, torch.device]) -> torch.device:
+def canonicalize_device(device: DeviceLikeType) -> torch.device:
```
Cool

mruberry reviewed (commented):
Fixing the clone typo and changing `wrap_device` to consistently be `canonicalize_device` looks great -- but see comments line for other changes

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

Lezcano commented:
Fair enough, I removed the sugar (which was inspired by `t.is_cpu()` and `t.is_cuda()` we currently have in PyTorch). Now this is just a general clean-up PR.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1995,7 +1995,7 @@ def tensor_split(
     # If indices_or_sections is a tensor, it must be a CPU Long tensor
     if isinstance(indices_or_sections, TensorLike):
-        if indices_or_sections.device != torch.device("cpu"):
+        if not indices_or_sections.device.type == "cpu":
```
Nice fix

mruberry reviewed (approved):
Cool!

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]

lezcano committed:
Update on "Fix handling of device"


Removes an unnecessary auxiliary function (we had already implemented
it), uses DeviceLikeType to denote str or dtype, and adds `is_cpu` and
`is_cuda` helper functions

[ghstack-poisoned]
